### ISSUE TITLE
Retrying is not working on Guzzle 7

### ISSUE THREAD

**Issue by hareku** [algolia/algoliasearch-client-php#640]
- Algolia Client Version: 2.7.0
- Language Version: PHP 7.4

### Description
Guzzle 7 has breaking changes at the structure of exceptions.
https://github.com/guzzle/guzzle/tree/7.0/src/Exception

`Guzzle6HttpClient` catches `GuzzleHttp\Exception\RequestException`, but Guzzle 7 throws `GuzzleHttp\Exception\ConnectException` if timeout occured.

https://github.com/algolia/algoliasearch-client-php/blob/2.7.0/src/Http/Guzzle6HttpClient.php#L28

**Comment by e1himself** [comment]
@hareku is Guzzle 7 even supported? 

Judging by the `Guzzle6HttpClient` name I conclude it's tailored to Guzzle 6.

Even though the Travis tests matrix has Guzzle 7 as one of its dimentions, it seems that only Guzzle 6 and older versions are being used inside `Algolia::getHttpClient()`:

https://github.com/algolia/algoliasearch-client-php/blob/0342b0b8e3409c8a246cc6a223a8e8679f3e816d/.travis.yml#L65-L68
https://github.com/algolia/algoliasearch-client-php/blob/0342b0b8e3409c8a246cc6a223a8e8679f3e816d/src/Algolia.php#L98-L104

I guess the best way forward would be to submit a PR:
- introducing `Guzzle7HttpClient` next to the existing `Guzzle6HttpClient` adapter
- extending `Algolia::getHttpClient()` to instantiate it automatically when possible

**Comment by e1himself** [comment]
~Even though the Travis tests matrix has Guzzle 7 as one of its dimentions, it seems that only Guzzle 6 and older versions are being used inside `Algolia::getHttpClient()`:~

It seems it is indeed a bug. Either `Guzzle6HttpClient` should be fine working with Guzzle 7, or `Algolia::getHttpClient()` should not use `Guzzle6HttpClient` when Guzzle 7 is installed.

**Comment by e1himself** [comment]
It would be less of a problem if `Guzzle6HttpClient` was a different Composer package that would require Guzzle v6 explicitly. 

Example *composer.json*:
```
{
   "name": "algolia/algoliasearch-client-php-guzzle6-adapter",
   "require": {
       "algolia/algoliasearch-client-php": "2.*",
       "guzzlehttp/guzzle": "6.*"
    }
```

With this setup it's just not possible to use `Guzzle7HttpClient` with Guzzle of any version other than 6. 

Additionally `algolia/algoliasearch-client-php-guzzle7-adapter` can be implemented. Or any other.

**Comment by chloelbn** [comment]
Guzzle 7 is now supported on the PHP client, please let us know if you're still facing this issue and reopen the ticket!

**Comment by reieRMeister** [comment]
Closed without merging?

**Comment by chloelbn** [comment]
Hi @reieRMeister, it’s been added in a different PR

**Comment by thomasbley** [comment]
Retrying is still not working with Guzzle 7.2.0 and algolia/algoliasearch-client-php 2.7.3.

**Comment by greg0ire** [comment]
@chloelbn I'm having an issue, but I have more details I think:

https://github.com/algolia/algoliasearch-client-php/blob/58ad00a6580ad5c32d481c5544cc496291ee8a3e/src/Http/GuzzleHttpClient.php#L30

This line wrongly assumes `$e` has a `hasResponse()` method. But when you fail to connect, there is no response, which mean `ConnectException` does not have that method.

**Comment by chloelbn** [comment]
Fixed with #675
